clippy_lints/src/functions/impl_trait_in_params.rs

```diff
@@ -0,0 +1,88 @@
+use clippy_utils::{diagnostics::span_lint_and_then, is_in_test_function};
+
+use rustc_hir::{intravisit::FnKind, Body, Generics, HirId};
+use rustc_lint::LateContext;
+use rustc_span::Span;
+
+use super::IMPL_TRAIT_IN_PARAMS;
+
+pub(super) fn check_fn<'tcx>(cx: &LateContext<'_>, kind: &'tcx FnKind<'_>, body: &'tcx Body<'_>, hir_id: HirId) {
+    if cx.tcx.visibility(cx.tcx.hir().body_owner_def_id(body.id())).is_public() && !is_in_test_function(cx.tcx, hir_id)
+    {
+        if let FnKind::ItemFn(ident, generics, _) = kind {
+            for param in generics.params {
+                if param.is_impl_trait()
+                    && !param.name.ident().as_str().contains('<')
+                    && !param.name.ident().as_str().contains('(')
+                {
+                    // No generics with nested generics, and no generics like FnMut(x)
+                    span_lint_and_then(
+                        cx,
+                        IMPL_TRAIT_IN_PARAMS,
+                        param.span,
+                        &format!("'{}' in the function's parameters", param.name.ident().as_str()),
+                        |diag| {
+                            let next_letter = next_valid_letter(generics);
+                            if let Some(gen_span) = generics.span_for_param_suggestion() {
+                                diag.span_suggestion_with_style(
+                                    gen_span,
+                                    format!(
+                                        "create a generic type here and replace that `{}` with `{}`",
+                                        param.name.ident().as_str(),
+                                        next_letter
+                                    ),
+                                    ", T: Trait",
```

Comment:
This should use something like 'add a type parameter' as the help message. The suggestion should have the correct replacement (e.g. `format!(", {name}: {trait_name}")`).

Reply:
Fixed